### PR TITLE
Code gen reorganizing

### PR DESCRIPTION
**Pull request by WardBrian**
As discussed in #1146, the function code generation has gotten really messy. This PR doesn't fix that, but it does move it into it's own file behind a _very_ simple interface:

```ocaml
val pp_functions_functors : Format.formatter -> Program.Numbered.t -> unit
(** Pretty-print all user defined functions. Creates functor structs as needed
  *)

val pp_standalone_fun_def :
  string -> Format.formatter -> 'a Program.fun_def -> unit
(** Creates functions outside the model namespaces which only call the ones
   inside the namespaces *)
```

This PR changes no code, just where it is located. Hopefully it will make it easier for @SteveBronder or I to tackle the larger refactoring in the near future. 

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] OR, no user-facing changes were made

## Release notes

Reorganize code gen modules

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/stan_math_backend/Function_gen.ml [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+       ; return: string
+       ; body: stmt list option }
+   | Generated of string
+       (** Placeholder for all other C++.
+           We should slowly move more and more statements
+           away from strings *) *)
+
+let pp_template ppf template =
+  match template with
+  | Typename t -> pf ppf "typename %s" t
```
[the remark this answers is not in the document]
Comment (WardBrian) [reply src/stan_math_backend/Function_gen.ml:30]: I don’t, though would that make more sense as part of a refactor or an attempt to clean up how the require templates are working?

As soon as the patch is out I plan to merge #1133, so the require stuff would be great to tackle next

File: test/unit/Stan_math_code_gen_tests.ml [hunk 2]
```diff
@@ -1,113 +1 @@
-open Middle
-open Stan_math_backend
```
Comment (SteveBronder) [inline test/unit/Stan_math_code_gen_tests.ml:2]: Why is this file deleted?

**Comment by WardBrian** [comment]
@rok-cesnovar just as an FYI, this is fine if it gets merged before the patch, it's just cut/pasting code into different files for organizational reasons and doesn't need a release note.

**Comment by rok-cesnovar** [comment]
Merge away!
